[36384] Add job state time limit actions to batch queue

jchorl opened the pull request:
### Description
A new option has been added to batch job queues: https://aws.amazon.com/about-aws/whats-new/2024/03/aws-batch-alerts-detect-jobs-runnable-state/

This PR just adds support for the new API fields to the batch queue terraform resource.


### Relations

Closes #36384

### References

API reference: https://docs.aws.amazon.com/batch/latest/APIReference/API_JobStateTimeLimitAction.html


### Output from Acceptance Testing

I didn't run acceptance tests to avoid dangling resources in my account.

I did however use this to update job queues and it worked.

drewmullen commented on internal/service/batch/job_queue.go:
```diff
@@ -133,6 +133,25 @@ func (r *resourceJobQueue) Schema(ctx context.Context, request resource.SchemaRe
+				Attributes: map[string]schema.Attribute{
+					"action": schema.StringAttribute{
+						Required: true,
+					},
```
```suggestion
						ValidateFunc: validation.StringInSlice(ec2.JobStateTimeLimitActionsAction_Values(), false),
					},
```

drewmullen commented on internal/service/batch/job_queue.go:
```diff
@@ -133,6 +133,25 @@ func (r *resourceJobQueue) Schema(ctx context.Context, request resource.SchemaRe
+					},
+					"max_time_seconds": schema.Int64Attribute{
+						Required: true,
+					},
```
```suggestion
						ValidateFunc: validation.IntBetween(600, 86400),
					},
```

drewmullen commented on internal/service/batch/job_queue.go:
```diff
@@ -133,6 +133,25 @@ func (r *resourceJobQueue) Schema(ctx context.Context, request resource.SchemaRe
+					},
+					"state": schema.StringAttribute{
+						Required: true,
+					},
```
```suggestion
						ValidateFunc: validation.StringInSlice(ec2.JobStateTimeLimitActionsState_Values(), false),
					},
```

drewmullen commented on internal/service/batch/job_queue_test.go:
```diff
@@ -453,6 +453,38 @@ func TestAccBatchJobQueue_tags(t *testing.T) {
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
```
- [ ] Please add a test with 2 `job_state_time_limit_action` defined 
- [ ] Please add a test where you update the `reason` and `max_time_seconds`

jchorl replied:
Done!

drewmullen commented on website/docs/r/batch_job_queue.html.markdown:
```diff
@@ -86,6 +87,13 @@ This resource supports the following arguments:
+### job_state_time_limit_action
+
+* `action` - (Required) The action to take when a job is at the head of the job queue in the specified state for the specified period of time.
+* `max_time_seconds` - (Required) The approximate amount of time, in seconds, that must pass with the job in the specified state before the action is taken.
```
```suggestion
    * `job_state_time_limit_action.#.max_time_seconds` - The approximate amount of time, in seconds, that must pass with the job in the specified state before the action is taken. Valid values include integers between `600` & `86400`
```

drewmullen commented on website/docs/r/batch_job_queue.html.markdown:
```diff
@@ -86,6 +87,13 @@ This resource supports the following arguments:
+* `action` - (Required) The action to take when a job is at the head of the job queue in the specified state for the specified period of time.
+* `max_time_seconds` - (Required) The approximate amount of time, in seconds, that must pass with the job in the specified state before the action is taken.
+* `reason` - (Required) The reason to log for the action being taken.
+* `state` - (Required) The state of the job needed to trigger the action.
```
```suggestion
* `state` - (Required) The state of the job needed to trigger the action. Valid values include `"RUNNABLE"`.
```

drewmullen commented on website/docs/r/batch_job_queue.html.markdown:
```diff
@@ -86,6 +87,13 @@ This resource supports the following arguments:
 
+### job_state_time_limit_action
+
+* `action` - (Required) The action to take when a job is at the head of the job queue in the specified state for the specified period of time.
```
```suggestion
* `action` - (Required) The action to take when a job is at the head of the job queue in the specified state for the specified period of time. Valid values include `"CANCEL"`
```

drewmullen reviewed (changes requested):
Thanks for opening this PR. Was planning to get to that this week... appreciate you stepping up. Here's some requested changes:

Also please include a [changelog entry](https://hashicorp.github.io/terraform-provider-aws/changelog-process/)

drewmullen commented on internal/service/batch/job_queue.go:
```diff
@@ -259,6 +292,11 @@ func (r *resourceJobQueue) Update(ctx context.Context, request resource.UpdateRe
+	if !plan.JobStateTimeLimitAction.IsNull() && !plan.JobStateTimeLimitAction.Equal(state.JobStateTimeLimitAction) {
+		flex.Expand(ctx, plan.JobStateTimeLimitAction, &input.JobStateTimeLimitActions)
+		update = true
+	}
```
can `reason` be updated?

drewmullen commented:
@jchorl looks like I gave you the validation functions using the old framework and this resource is on the new framework… I can send some examples tomorrow if you have trouble

jchorl commented:
> @jchorl looks like I gave you the validation functions using the old framework and this resource is on the new framework… I can send some examples tomorrow if you have trouble

You're too fast, just figuring it out locally. Will push updates soon.

drewmullen commented:
I saw the notification and immediately felt bad for giving bad code lol

@jchorl

jchorl commented:
I think I addressed all feedback. Unfortunately I can't test updating the `reason` this sec because there's some unrelated concurrent changes in-flux that can't be targetted around. I can try to test tomorrow.

drewmullen commented on internal/service/batch/job_queue_test.go:
```diff
@@ -453,6 +453,103 @@ func TestAccBatchJobQueue_tags(t *testing.T) {
+			},
+			{
+				Config: acctest.ConfigCompose(
+					testAccJobQueueConfigBase(rName),
```
```suggestion
					testAccJobQueueConfig_Base(rName),
```

ChannyClaus commented:
@jchorl are you still working on this PR? if you're busy with other things, happy to take over (this feature would be incredibly helpful for us 😅)

jchorl commented:
> @jchorl are you still working on this PR? if you're busy with other things, happy to take over (this feature would be incredibly helpful for us 😅)

Please feel free to take it over - I haven't had a chance to figure out terraform tests, especially with all the resources they configure in your AWS acct.

I do think there is a lingering issue in this PR with state getting out of sync - i.e. if you `apply` the same config twice I think it still shows changes.

ChannyClaus commented:
> > @jchorl are you still working on this PR? if you're busy with other things, happy to take over (this feature would be incredibly helpful for us 😅)
> 
> Please feel free to take it over - I haven't had a chance to figure out terraform tests, especially with all the resources they configure in your AWS acct.
> 
> I do think there is a lingering issue in this PR with state getting out of sync - i.e. if you `apply` the same config twice I think it still shows changes.

made a PR https://github.com/hashicorp/terraform-provider-aws/pull/38784 to pick this up. do you happen to remember the configuration with which you ran into the out-of-sync state issue? seems like when i make changes to the `job_state_time_limit_action` attribute, the subsequent runs of `terraform apply` shows no change.

on a related-ish note, hopefully someone will be able to take a look at the PR soon, since given the acquisition going on at my company, it's unclear how much longer i'll have access to AWS account here :.)

jchorl commented:
> > > @jchorl are you still working on this PR? if you're busy with other things, happy to take over (this feature would be incredibly helpful for us 😅)
> > 
> > 
> > Please feel free to take it over - I haven't had a chance to figure out terraform tests, especially with all the resources they configure in your AWS acct.
> > I do think there is a lingering issue in this PR with state getting out of sync - i.e. if you `apply` the same config twice I think it still shows changes.
> 
> made a PR #38784 to pick this up. do you happen to remember the configuration with which you ran into the out-of-sync state issue? seems like when i make changes to the `job_state_time_limit_action` attribute, the subsequent runs of `terraform apply` shows no change.
> 
> on a related-ish note, hopefully someone will be able to take a look at the PR soon, since given the acquisition going on at my company, it's unclear how much longer i'll have access to AWS account here :.)

I just built my branch and tf-applied off that provider. Got a bunch of:
```
│ Error: Provider produced inconsistent final plan                                                                                                                                             
│                                                                                                                                                                                              
│ When expanding the plan for module.whatever.compute_env["foo"] to include new values learned so far during apply, provider               
│ "registry.terraform.io/hashicorp/aws" changed the planned action from Update to DeleteThenCreate.                                                                                            
│                                                                                                                                                                                              
│ This is a bug in the provider, which should be reported in the provider's own issue tracker. 
```

I'm not sure what this error means/why it would occur. This was off just a normal `terraform apply`. If you can toggle on/off the `job_state_time_limit_action`, tf apply, and not hit this, maybe it's fixed.

Genuine thanks for picking this up.

ChannyClaus commented:
yup, tried it with 
```
terraform {
  required_providers {
	aws = {
      source  = "terraform.local/local/aws"
    }
  }
}

# Configure the AWS Provider
provider "aws" {
  region = "us-west-2"
}


resource "aws_batch_compute_environment" "sample" {
  compute_environment_name = "sample"

  compute_resources {
    max_vcpus = 16
    subnets = [
      "<redacted>",
    ] 
    security_group_ids = [
      "<redacted>",
    ]
    type = "FARGATE"
  }

  type         = "MANAGED"
}

resource "aws_batch_job_queue" "test_queue" {
  name     = "tf-test-batch-job-queue4"
  state    = "ENABLED"
  priority = 1
  compute_environment_order {
    order               = 1
    compute_environment = aws_batch_compute_environment.sample.arn
  }
  job_state_time_limit_action {
    action           = "CANCEL"
    max_time_seconds = 606
    reason           = "CAPACITY:INSUFFICIENT_INSTANCE_CAPACITY"
    state            = "RUNNABLE"
  }
}
```
i don't seem to get that error message.

and no problem! we ended up setting up a cronjob to pick up on jobs that get stuck on RUNNABLE and i'm just glad AWS got around to support this natively on their end.
